trin-core/src/portalnet/find/query_pool.rs

```diff
@@ -0,0 +1,295 @@
+    /// The instant when the query started (i.e. began waiting for the first
+    /// result from a peer).
+    started: Option<Instant>,
+    /// Target we are looking for.
+    target: TTarget,
+    /// Phantom use of TResult generic type, which is used in implementation.
+    unused_result: PhantomData<TResult>,
+    
+}
+
+/// The peer selection strategies that can be used by queries.
+enum QueryPeerIter<TNodeId> {
+    FindNode(FindNodeQuery<TNodeId>)
+}
+
+impl<TTarget, TNodeId, TResult> Query<TTarget, TNodeId, TResult>
+where
+    TTarget: TargetKey<TNodeId>,
+    TNodeId: Into<Key<TNodeId>> + Eq + Clone,
+    TResult: Into<TNodeId> + Clone,
+{
+    /// Creates a new query without starting it.
+    fn new(id: QueryId, peer_iter: QueryPeerIter<TNodeId>, target: TTarget) -> Self {
+        Query {
+            id,
+            peer_iter,
+            target,
+            started: None,
+            unused_result: PhantomData,
+        }
+    }
+
+    /// Gets the unique ID of the query.
+    pub fn id(&self) -> QueryId {
+        self.id
+    }
+
+    /// Informs the query that the attempt to contact `peer` failed.
+    pub fn on_failure(&mut self, peer: &TNodeId) {
+        match &mut self.peer_iter {
+            QueryPeerIter::FindNode(iter) => iter.on_failure(&peer),
+        }
+    }
+
+    /// Informs the query that the attempt to contact `peer` succeeded,
+    /// possibly resulting in new peers that should be incorporated into
+    /// the query, if applicable.
+    pub fn on_success<'a>(&mut self, peer: &TNodeId, new_peers: &'a [TResult])
+    where
+        &'a TResult: Into<TNodeId>,
+    {
+        match &mut self.peer_iter {
+            QueryPeerIter::FindNode(iter) => {
+                iter.on_success(peer, new_peers.iter().map(|result| result.into()).collect())
+            }
+        }
+    }
+
+    /// Advances the state of the underlying peer iterator.
+    fn next(&mut self, now: Instant) -> QueryState<TNodeId> {
+        match &mut self.peer_iter {
+            QueryPeerIter::FindNode(iter) => iter.next(now),
+        }
+    }
+
+    /// Consumes the query, producing the final `QueryResult`.
+    pub fn into_result(self) -> QueryResult<TTarget, impl Iterator<Item = TNodeId>> {
+        let peers = match self.peer_iter {
+            QueryPeerIter::FindNode(iter) => iter.into_result(),
+        };
+        QueryResult {
+            target: self.target,
+            closest_peers: peers.into_iter(),
+        }
+    }
+
+    /// Returns a reference to the query `target`.
+    pub fn target(&self) -> &TTarget {
+        &self.target
+    }
+
+    /// Returns a mutable reference to the query `target`.
+    pub fn target_mut(&mut self) -> &mut TTarget {
+        &mut self.target
+    }
+}
+
+/// The result of a `Query`.
+pub struct QueryResult<TTarget, TClosest> {
+    /// The target of the query.
+    pub target: TTarget,
+    /// The closest peers to the target found by the query.
+    pub closest_peers: TClosest,
+}
```

Comment:
Have you considered how we would extend this to results for `FindContent` queries, where the result may not contain a list of closest peers?